source/server/listener_impl.cc

```diff
@@ -59,42 +93,26 @@ Network::SocketSharedPtr ListenSocketFactoryImplBase::createListenSocketAndApply
   return socket;
 }
 
-void ListenSocketFactoryImplBase::setLocalAddress(
-    Network::Address::InstanceConstSharedPtr local_address) {
-  ENVOY_LOG(debug, "Set listener {} socket factory local address to {}", listener_name_,
-            local_address->asString());
-  local_address_ = local_address;
-}
+Network::SocketSharedPtr ListenSocketFactoryImpl::getListenSocket() {
 
-TcpListenSocketFactory::TcpListenSocketFactory(
-    ListenerComponentFactory& factory, Network::Address::InstanceConstSharedPtr local_address,
-    const Network::Socket::OptionsSharedPtr& options, bool bind_to_port,
-    const std::string& listener_name)
-    : ListenSocketFactoryImplBase(factory, local_address, Network::Address::SocketType::Stream,
-                                  options, bind_to_port, listener_name) {
-  socket_ = createListenSocketAndApplyOptions();
-  if (socket_ != nullptr && localAddress()->ip() != nullptr && localAddress()->ip()->port() == 0) {
-    setLocalAddress(socket_->localAddress());
-  }
-}
+  if (reuse_port_) {
+    Network::SocketSharedPtr socket;
+    absl::call_once(steal_once_, [this, &socket]() {
```

Comment:
TBH I find this logic to be a little bit confusing. How about we just always store a copy of the socket on the listener config regardless of whether it is reuse port or not, port 0 or not, etc. Then we only need to store a bool of whether it's reuse port or not so the socket should be recreated? As long as we never listen on the socket on the main thread I don't think it will matter if it's sitting there? 

I think this would unify the logic quite a bit. WDYT?

Reply:
Would it be more clear if we separate these implementations into different factories, I.e. TcpSharedSocketFactory, TcpReusePortSocketFactory, and UdpSocketFactory? TcpSharedSocketFactory is just as the original implementation. TcpReusePortSocketFactory can be implemented with a temporary socket which is transferred to listeners later. UdpSocketFactory should needs more care about socket life time as we want to keep all the sockets open till all listeners stop  listening.

Reply:
IMHO, adding factories for this purpose makes things more complicated.

Reply:
> For UDP, if a bound socket is not used, kernel still forwards packets to it, but no worker thread handles them, so its socket fd must be closed.

OK, right, that makes sense. Can you add more comments about why we do it this way? It's not clear from the comment trail all of the discussion that was done on this topic and we should make sure it's not lost.

> IMHO, adding factories for this purpose makes things more complicated.

I don't have a strong opinion on this so will defer to @danzh2010. I think we can fix it up later if needed.